Sources/KituraWebSocket/PermessageDeflateInflater.swift

```diff
@@ -0,0 +1,149 @@
+class PermessageDeflateInflater : ChannelInboundHandler {
+    typealias InboundIn = WebSocketFrame 
+    typealias InboundOut = WebSocketFrame
+
+    public init(maxWindowBits: Int32 = 15, noContextTakeOver: Bool = false) {
+        self.maxWindowBits = maxWindowBits
+        self.noContextTakeOver = noContextTakeOver
+    }
+
+    var noContextTakeOver: Bool
+
+    /// The zlib stream
+    private var stream: z_stream = z_stream()
+
+    /// A buffer to accumulate payload across multiple frames
+    var payload: ByteBuffer?
+
+    /// Is this a text or binary message? Continuation frames don't have this information.
+    private var messageType: WebSocketOpcode?
+
+    /// The default LZ77 window size; 15
+    var maxWindowBits = MAX_WBITS
+
+    /// PermessageDeflateInflater is a `ChannelInboundHandler`, this function gets called when the previous inbound handler fires a channel read event.
+    /// Here, we intercept incoming compressed frames, decompress the payload across multiple continuation frame and write a fire a channel read event
+    /// with the entire frame data decompressed.
+    func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+        var frame = unwrapInboundIn(data)
+        // If this is a control frame, there's nothing to do.
+        guard frame.opcode == .text || frame.opcode == .binary || frame.opcode == .continuation else {
+            ctx.fireChannelRead(self.wrapInboundOut(frame))
+            return
+        }
+
+        // If this is a continuation frame, have the payload appended to `payload`, else set `payload` and store the messageType
+        var receivedPayload = frame.unmaskedData
+        if frame.opcode == .continuation {
+            self.payload?.write(buffer: &receivedPayload)
+        } else {
+            self.messageType = frame.opcode
+            self.payload = receivedPayload
+        }
+
+        // If the current frame isn't a final frame of a message or if `payload` still empty, there's nothing to do.
+        guard frame.fin, var inputBuffer = self.payload else { return }
+
+        // Append the trailer 0, 0, ff, ff before decompressing
+        inputBuffer.write(bytes: [0x00, 0x00, 0xff, 0xff])
+        var inflatedPayload = inflatePayload(in: inputBuffer, allocator: ctx.channel.allocator)
+
+        // Apply the WebSocket mask on the inflated payload
+        inflatedPayload.webSocketMask(frame.maskKey!)
+
+        // Create a new frame with the inflated payload and pass it on to the next inbound handler, mostly WebSocketConnection
+        let inflatedFrame = WebSocketFrame(fin: true, rsv1: false, opcode: self.messageType!, maskKey: frame.maskKey!, data: inflatedPayload)
+        ctx.fireChannelRead(self.wrapInboundOut(inflatedFrame))
+    }
+
+    func inflatePayload(in buffer: ByteBuffer, allocator: ByteBufferAllocator) -> ByteBuffer {
```

Comment:
I don't think we gain much from having initializeDecoder and deinitializeDecoder as separate functions since they are only called once here. Might be easier to follow if you combine these into one function.

Reply:
Fair point. Done!